File: aq/src/core/paths.ts
import { homedir } from "node:os"
import path from "node:path"
import { readRecipeFile } from "./recipe-yaml.js"

export function insideTrain(train: string, rel: string): string {
  const root = path.resolve(train)
  const dest = path.resolve(root, rel)
  const relTo = path.relative(root, dest)
  if (relTo.startsWith("..") || path.isAbsolute(relTo)) {
    throw new Error(`outside train: ${rel}`)
  }
  return dest
}

export function shortPath(p: string): string {
  const abs = path.resolve(p)
  const home = homedir()
  if (abs === home) return "~"
  if (abs.startsWith(home + path.sep)) return "~" + abs.slice(home.length)
  return abs
}

/** Defaults match the classic layout; override via recipe.yaml `paths:`. */
export const PATH_DEFAULTS = {
  artifacts: "artifacts",
  evals: "evals",
} as const

export type PathKey = keyof typeof PATH_DEFAULTS

export function pathMap(trainOrRecipe: string): Record<PathKey, string> {
  const rec = readRecipeFile(trainOrRecipe)
  const raw = (rec.paths && typeof rec.paths === "object" && !Array.isArray(rec.paths)
    ? (rec.paths as Record<string, unknown>)
    : {}) as Record<string, unknown>
  const out: Record<PathKey, string> = { ...PATH_DEFAULTS }
  for (const key of Object.keys(PATH_DEFAULTS) as PathKey[]) {
    const v = raw[key]
    if (v == null || v === false) continue
    const s = String(v).trim()
    if (s) out[key] = s
  }
  return out
}

export function pathRel(trainOrRecipe: string, key: PathKey): string {
  return pathMap(trainOrRecipe)[key]
}

/** Resolve a layout path (relative to train root, or absolute). */
export function resolveLayoutPath(train: string, relOrAbs: string): string {
  if (path.isAbsolute(relOrAbs)) return path.resolve(relOrAbs)
  return path.resolve(train, relOrAbs)
}

export function artifactsDir(train: string): string {
  return resolveLayoutPath(train, pathRel(train, "artifacts"))
}

export function evalsDir(train: string): string {
  return resolveLayoutPath(train, pathRel(train, "evals"))
}

File: aq/kernel/protocol/paths.py
"""Layout paths from recipe.yaml — defaults match the classic tree.

    paths:
      artifacts: artifacts   # runtime root (checkpoints, metrics, plots, …)
      evals: evals           # probe set directory

Relative paths are under the train root. Absolute paths are allowed.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any

DEFAULTS: dict[str, str] = {
    "artifacts": "artifacts",
    "evals": "evals",
}


def peek_recipe(train: Path) -> dict[str, Any]:
    """Load recipe.yaml without full validation (for path resolution)."""
    recipe = train / "recipe.yaml"
    if not recipe.is_file():
        return {}
    try:
        from protocol.recipe import parse_recipe

        return parse_recipe(recipe)
    except Exception:
        return {}


def path_map(rec: dict[str, Any] | None) -> dict[str, str]:
    """Resolved path map (defaults filled in)."""
    out = dict(DEFAULTS)
    raw = (rec or {}).get("paths") if isinstance(rec, dict) else None
    if isinstance(raw, dict):
        for key in DEFAULTS:
            val = raw.get(key)
            if val is None or val is False:
                continue
            s = str(val).strip()
            if s:
                out[key] = s
    return out


def rel(rec: dict[str, Any] | None, key: str) -> str:
    if key not in DEFAULTS:
        raise KeyError(key)
    return path_map(rec)[key]


def resolve(train: Path, rel_or_abs: str) -> Path:
    p = Path(rel_or_abs).expanduser()
    if p.is_absolute():
        return p.resolve()
    return (train / p).resolve()


def art_dir(train: Path, rec: dict[str, Any] | None = None) -> Path:
    if rec is None:
        rec = peek_recipe(train)
    return resolve(train, rel(rec, "artifacts"))


def evals_dir(train: Path, rec: dict[str, Any] | None = None) -> Path:
    if rec is None:
        rec = peek_recipe(train)
    return resolve(train, rel(rec, "evals"))


def ckpt_dir(train: Path, rec: dict[str, Any] | None = None) -> Path:
    d = art_dir(train, rec) / "checkpoints"
    d.mkdir(parents=True, exist_ok=True)
    return d
